### PR TITLE
Core: Add `renderLabel` to customize sidebar tree labels

### PR DESCRIPTION
**Pull request by tooppaaa**
Issue: Lack of capability to add elements to

(Updated after @ghengeveld work)

sidebar subheadings (most important one in my opinion)
sidebar items
folder
component
story
This PR aims as a proposal for this and basis of discussions

image

Should we keep double icons when prefix is set ?
Should it be called prefix ?

### CODE REVIEW

**Comment by shilman** [comment]
cc @domyen @ghengeveld let's see what we can do to get this over the line

**Comment by wKich** [comment]
Yeah. It would be great to add prefix for items.

**Comment by saulmases** [comment]
I comment here because there is also interest in being able to customise storybook sidebar, like adding a prefix before the items so we can add there things like component status in the sidebar.

**Comment by wKich** [comment]
@tooppaaa, @shilman, can we also add a method to the client API that allows changing prefixes from code?

**Comment by tooppaaa** [comment]
@wKich, that can be an interesting feature. I'd like to get this one merged before bringing more changes

**Comment by tooppaaa** [comment]
@shilman merged next into this one.

**Comment by tooppaaa** [comment]
@ghengeveld any update ?

**Comment by ghengeveld** [comment]
Hey, I would like to propose the following:

- We rename `storyPrefix` to `sidebar` in the config.
- This accepts an object, with a `storyLabel` property (and potentially other sidebar config options, later down the road).
- `storyLabel` is a function, which takes an Item (`Root | Group | Story`) and should return a string or ReactNode.
- The default implementation would be `item => item.name`.

This would allow much more flexibility in terms of how the label is rendered in the sidebar.

**Comment by tooppaaa** [comment]
Done

File: lib/ui/src/components/sidebar/Tree.tsx [hunk 1]
```diff
@@ -91,7 +92,7 @@ const Node = React.memo<NodeProps>(
             onSelectStoryId(item.id);
           }}
         >
-          {item.name}
+          {item.storyLabel || item.name}
```
[the remark this answers is not in the document]
Comment (tooppaaa) [reply lib/ui/src/components/sidebar/Tree.tsx:95]: The user interface for this is a function: ```storyLabel: ({ id, name }) => {}```, internally, it's a react node

**Comment by tooppaaa** [comment]
@ghengeveld, I can't be keeping up to date with PR's that are open for that long time. Can we either get it moving or declined ?

File: lib/ui/src/components/sidebar/Tree.stories.tsx [hunk 2]
```diff
@@ -58,6 +59,7 @@ const singleStoryComponent = {
     isLeaf: true,
     isComponent: false,
     isRoot: false,
+    storyLabel: <span>🔥 Single</span>,
```
[the remark this answers is not in the document]
Comment (shilman) [reply lib/ui/src/components/sidebar/Tree.stories.tsx:62]: No, I think this is the output of the label function (e.g. the values in the map that he defines in his example)

**Review by shilman — CHANGES_REQUESTED** [review]
@ghengeveld This all looks good to me. The only thing that's weird about it is that the user specifies the `sideBar.storyLabel` function, and the result of that function is a react node called `storyLabel`, which is confusing. I think the function should be called `createStoryLabel` or something more explicit.

Also I thought about the portability of this feature and I think it's probably fine, so long as the user doesn't depend on the `storyLabel` for anything. If we consider it like "visual enhancement" to the underlying data, then other ecosystem tools can just ignore this configuration and provide a downgraded experience which should be fine.

**Comment by ghengeveld** [comment]
I brought this up-to-date with `next`, as well as upcoming changes from `sidebar-config`, so this now telescopes on `sidebar-config`.

Along with the merge I made some API changes and updated the docs:
- The config option is now `sidebar.renderLabel` which accepts a function `(item: Root | Group | Story) => ReactNode`
- Root, Group and Story now have a `label` property which contains the result of invoking `renderLabel`.

That should avoid any confusion over `storyLabel`.

**Review by shilman — APPROVED** [review]
Looking great @ghengeveld @tooppaaa 💯

**Pull request merged**